File: 2025/Heeho/2025-06-15-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int maxLen = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            while (set.contains(s.charAt(right))) {
                set.remove(s.charAt(left));
                left++;
            }

            set.add(s.charAt(right));
            maxLen = Math.max(maxLen, right - left + 1);
        }

        return maxLen;
    }
}

File: 2025-06-15-Longest_Substring_Without_Repeating_Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        save = set()
        max_len = 0
        left = 0
        for i in range(len(s)):
            while s[i] in save :
                save.remove(s[left])
                left += 1
            save.add(s[i])
            max_len = max(max_len, i - left + 1)
            print(save, max_len)
        return max_len
